File: services/memory-service/app/schemas/mcp.py
"""MCP server 管理 API 模型（M-09 扩展：热插拔）。

请求体直接复用 mcp_client.McpServerConfig 的字段约束（name 模式/transport
枚举/参数匹配校验），响应体附带运行态（connected/tools）。
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class McpServerCreate(BaseModel):
    """添加 MCP server 请求体（POST /api/mcp/servers）。

    字段约束与 McpServerConfig 对齐：name 限小写字母/数字/中划线，
    http 必填 url、stdio 必填 command（服务层校验）。
    """

    name: str = Field(
        min_length=1,
        max_length=64,
        pattern=r"^[a-z0-9][a-z0-9_-]{0,63}$",
        description="全局唯一标识（工具注册名前缀 mcp.<name>.*）",
    )
    transport: str = Field(pattern="^(http|stdio)$", description="http / stdio")
    url: str | None = Field(default=None, max_length=2048, description="http 端点")
    command: str | None = Field(default=None, max_length=512, description="stdio 启动命令")
    args: list[str] = Field(default_factory=list, description="stdio 命令参数")
    env: dict[str, str] | None = Field(default=None, description="stdio 额外环境变量")


class McpServerRead(BaseModel):
    """server 配置 + 运行态（GET /api/mcp/servers 条目）。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    transport: str
    url: str | None = None
    command: str | None = None
    args: list[str] = Field(default_factory=list)
    enabled: bool
    source: str = Field(description="env=配置种子（不可删）/ user=API 添加")
    created_at: datetime
    connected: bool = Field(default=False, description="当前长连接是否存活")
    tools: list[str] = Field(default_factory=list, description="已注册工具名清单")

File: frontend/src/api/mcp.ts
/**
 * MCP 服务管理 API（EchoDesk 前端，M-09 扩展：热插拔）。
 *
 * 对接 memory-service `/api/mcp/servers`：清单（配置 + 运行态）、
 * 热添加（服务端试连成功才落库）、热移除（断连 + 注销工具）。
 * 全局共享模块：无 workspace 维度，登录即可管理。
 */
import { api } from "./client";

/** MCP server 配置 + 运行态条目（GET /api/mcp/servers）。 */
export interface McpServerInfo {
  /** 配置行 ID。 */
  id: string;
  /** 全局唯一标识（工具名前缀 mcp.<name>.*）。 */
  name: string;
  /** 传输方式。 */
  transport: "http" | "stdio";
  /** http 端点（stdio 为 null）。 */
  url: string | null;
  /** stdio 启动命令（http 为 null）。 */
  command: string | null;
  /** stdio 命令参数。 */
  args: string[];
  /** 是否启用。 */
  enabled: boolean;
  /** env=配置种子（不可删）/ user=API 添加。 */
  source: "env" | "user";
  /** 创建时间。 */
  created_at: string;
  /** 当前长连接是否存活。 */
  connected: boolean;
  /** 已注册工具名清单。 */
  tools: string[];
}

/** 热添加请求体（POST /api/mcp/servers）。 */
export interface McpServerCreatePayload {
  /** 全局唯一标识（小写字母/数字/中划线）。 */
  name: string;
  /** 传输方式。 */
  transport: "http" | "stdio";
  /** http 端点（transport=http 必填）。 */
  url?: string;
  /** stdio 启动命令（transport=stdio 必填）。 */
  command?: string;
  /** stdio 命令参数。 */
  args?: string[];
}

/**
 * 列出全部 MCP server（配置 + 运行态）。
 *
 * @returns server 清单。
 */
export async function listMcpServers(): Promise<McpServerInfo[]> {
  const { data } = await api.get<McpServerInfo[]>("/api/mcp/servers");
  return data;
}

/**
 * 热添加 MCP server（服务端试连成功后落库，工具即刻可用）。
 *
 * @param payload - server 配置。
 * @returns 落库后的 server 条目。
 */
export async function addMcpServer(payload: McpServerCreatePayload): Promise<McpServerInfo> {
  const { data } = await api.post<McpServerInfo>("/api/mcp/servers", payload);
  return data;
}

/**
 * 热移除 MCP server（断连 + 注销 mcp.<name>.* 工具 + 删配置）。
 *
 * @param name - server 标识。
 */
export async function removeMcpServer(name: string): Promise<void> {
  await api.delete(`/api/mcp/servers/${encodeURIComponent(name)}`);
}
